quickwit-storage/src/split_creation.rs

```diff
@@ -0,0 +1,404 @@
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+use std::collections::HashMap;
+use std::fmt::Debug;
+use std::io::{self, Cursor, ErrorKind, SeekFrom, Write};
+use std::iter;
+use std::ops::Range;
+use std::path::{Path, PathBuf};
+
+use async_trait::async_trait;
+use futures::StreamExt;
+use quickwit_common::HOTCACHE_FILENAME;
+use rusoto_core::ByteStream;
+use tantivy::directory::OwnedBytes;
+use tokio::io::{AsyncReadExt, AsyncSeekExt};
+use tokio_util::io::ReaderStream;
+
+use crate::{BundleStorageFileOffsets, PutPayload};
+
+trait DataSource: tokio::io::AsyncRead + Send + Sync + Unpin {}
+impl<T: tokio::io::AsyncRead + Send + Sync + Unpin> DataSource for T {}
+
+/// Payload of a split which builds the split bundle and hotcache on the fly and streams it to the
+/// storage.
+#[derive(Clone)]
+pub struct SplitPayload {
+    /// Total split size
+    pub split_size: usize,
+    /// Files and their target range in the split bundle.
+    pub files_and_offsets: Vec<(PathBuf, Range<usize>)>,
+    /// bytes range of the footer (hotcache + bundle metadata)
+    pub footer_range: Range<usize>,
+    /// data of the footer, containng
+    /// [bundle metadata, bundle metadata len 8byte, hotcache, hotcache len 8 bytes]
+    pub footer_bytes: Vec<u8>,
+    /// The end offset of the data provider [File, .. File, SplitFooter]
+    pub block_offsets: Vec<Range<usize>>,
+}
+
+impl SplitPayload {
+    /// Get start and end block index for a range, where start of the Range is in the start block
+    /// and end of the range is in the end block.
+    fn get_start_and_end_block_index(&self, range: Range<usize>) -> io::Result<(usize, usize)> {
+        get_start_and_end_block_index(&self.block_offsets, range)
+    }
+
+    /// Get datasource for data block at index.
+    ///
+    /// We apply seek now because the returned DataSource trait does not allow it.
+    async fn get_datasource(
+        &self,
+        index: usize,
+        start_file_offset: u64,
+    ) -> io::Result<Box<dyn DataSource>> {
+        if index < self.files_and_offsets.len() {
+            let (path, _) = &self.files_and_offsets[index];
+            let mut ds = Box::new(tokio::fs::File::open(path).await?);
+            if start_file_offset != 0 {
+                ds.seek(SeekFrom::Start(start_file_offset)).await?;
+            }
+            Ok(ds)
+        } else {
+            let mut ds = Box::new(Cursor::new(self.footer_bytes.to_vec()));
+            if start_file_offset != 0 {
+                ds.seek(SeekFrom::Start(start_file_offset)).await?;
+            }
+            Ok(ds)
+        }
+    }
+}
+
+#[async_trait]
+impl PutPayload for SplitPayload {
+    async fn len(&self) -> io::Result<u64> {
+        Ok(self.split_size as u64)
+    }
+
+    async fn range_byte_stream(&self, range: Range<u64>) -> io::Result<ByteStream> {
+        let (start_index, end_index) =
+            self.get_start_and_end_block_index(range.start as usize..range.end as usize)?;
```

Comment:
In the same function, you use the word start and end for `[start, end)` AND `[start, end]` semantics.

Reply:
If you can remove the use of inclusive range this is better. If you cannot, please return  a RangeInclusive object.

Reply:
Ah so it is not doing exactly doing what I thought it was doing.

Reply:
Your `get_start_and_end_block_index` function doc is very clear. Then implementation is also pretty easy.
The contract choice however is rather unexpected. I don't think people would expect that.

There is a clunky side effect here. The last datasource can be empty. I don't think it yields a bug (I'd like a unit test about it though), but we end up opening a file for no reason at all.